lib/wasi/src/runtime/module_cache/types.rs

```diff
@@ -0,0 +1,164 @@
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    ops::Deref,
+    path::PathBuf,
+};
+
+use sha2::{Digest, Sha256};
+use wasmer::{Engine, Module};
+
+use crate::runtime::module_cache::AndThen;
+
+/// A cache for compiled WebAssembly modules.
+///
+/// ## Assumptions
+///
+/// Implementations can assume that cache keys are unique and that using the
+/// same key to load or save will always result in the "same" module.
+///
+/// Implementations can also assume that [`ModuleCache::load()`] will
+/// be called more often than [`ModuleCache::save()`] and optimise
+/// their caching strategy accordingly.
+#[async_trait::async_trait]
+pub trait ModuleCache: Debug {
+    async fn load(&self, key: Key, engine: &Engine) -> Result<Module, CacheError>;
+
+    async fn save(&self, key: Key, module: &Module) -> Result<(), CacheError>;
+
+    /// Chain a second cache onto this one.
+    ///
+    /// The general assumption is that each subsequent cache in the chain will
+    /// be significantly slower than the previous one.
+    ///
+    /// ```rust
+    /// use wasmer_wasix::runtime::module_cache::{
+    ///     ModuleCache, ThreadLocalCache, FileSystemCache, SharedCache,
+    /// };
+    ///
+    /// let cache = SharedCache::default()
+    ///     .and_then(FileSystemCache::new("~/.local/cache"));
+    /// ```
+    fn and_then<C>(self, other: C) -> AndThen<Self, C>
+    where
+        Self: Sized,
+        C: ModuleCache,
+    {
+        AndThen::new(self, other)
+    }
+}
+
+#[async_trait::async_trait]
+impl<D, C> ModuleCache for D
+where
+    D: Deref<Target = C> + Debug + Send + Sync,
+    C: ModuleCache + Send + Sync + ?Sized,
+{
+    async fn load(&self, key: Key, engine: &Engine) -> Result<Module, CacheError> {
+        (**self).load(key, engine).await
+    }
+
+    async fn save(&self, key: Key, module: &Module) -> Result<(), CacheError> {
+        (**self).save(key, module).await
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum CacheError {
+    #[error("Unable to serialize the module")]
+    Serialize(#[from] wasmer::SerializeError),
+    #[error("Unable to deserialize the module")]
+    Deserialize(#[from] wasmer::DeserializeError),
+    #[error("Unable to read from \"{}\"", path.display())]
+    Read {
+        path: PathBuf,
```

Comment:
The cache might not be able to read, and there might be no path (for example, if you are using a cache saved in redis via a key).
We should remove the path field here as is filesystem dependant

Reply:
See my response from your comment on the `Write` variant. TL;DR: These variants are for specific IO errors that may happen when you do filesystem operations, rather than a general *"some sort of read operation failed"* error.